src/main/java/com/opentable/kafka/util/OffsetMetrics.java

```diff
@@ -0,0 +1,242 @@
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Reservoir;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.opentable.concurrent.OTExecutors;
+import com.opentable.metrics.AtomicLongGauge;
+import com.opentable.metrics.graphite.MetricSets;
+
+/**
+ * Class that monitors consumer groups consuming from topics; registers its metrics with a metric registry.
+ *
+ * Runs a thread that polls the Kafka broker; use {@link #start()} and {@link #stop()} to spin up and down. They are
+ * annotated with post-construct and pre-destroy, respectively, in order to simplify Spring integration.
+ *
+ * The metric namespace is as follows; each metric is a {@link com.codahale.metrics.Gauge<Long>}.
+ *
+ * <p>
+ * {@code <metric prefix>.<topic>.{partition.%d,total}.{size,offset,lag}}
+ *
+ * <p>
+ * Size is the maximum offset of the topic (this is independent of any consumer). Offset is the current offset of the
+ * consumer. Lag is the size minus the offset. Totals are summed across the partitions.
+ * NB: Since the queries for the topic sizes and consumer offsets are necessarily separate, they race; this means that
+ * lag can possibly be negative.
+ *
+ * <p>
+ * In addition to these gauges, lag distribution {@link Histogram}s are also available as follows.
+ *
+ * <p>
+ * {@code <metric prefix>.<topic>.total.lag-distribution}
+ *
+ * <p>
+ * Future work: Have alternate constructor in which you don't specify any topics, and the class uses the
+ * {@link OffsetMonitor} to dynamically register metrics based on the topics that the consumer group is consuming.
+ *
+ * <p>
+ * Future work: Have alternate constructor in which you don't specify any groups either, and the class uses the
+ * {@link OffsetMonitor} to dynamically register metrics based on all consumer groups' topic consumption.
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+public class OffsetMetrics implements Closeable {
+    private static final Logger LOG = LoggerFactory.getLogger(OffsetMetrics.class);
+
+    private final String metricPrefix;
+    private final MetricRegistry metricRegistry;
+    private final String groupId;
+    private final Collection<String> topics;
+    private final Duration pollPeriod;
+    private final OffsetMonitor monitor;
+    private final Map<String, Metric> metricMap;
+    private final ScheduledExecutorService exec;
+
+    public static OffsetMetricsBuilder builder(
+            final String metricPrefix,
```

Comment:
It's a little strange to have so many required parameters to make a builder.  Especially given that three of them are `String` typed arguments, might it be clearer to have them provided as builder methods and then assert that they are set on `build()`?

Reply:
It might be a little strange, but the alternative you suggest just seems so crazy to me.  The arguments to `builder` here are all of the required ones.  We literally have the ability with this statically-typed language to make sure they get passed.  Why would we defer detection of bugs to run time?